File: 高级数据结构与算法/Java/第3章 排序算法/3.3 归并排序/MergeSortIteration.java
import java.util.Arrays;

public class MergeSortIteration {
    public static void main(String[] args) {
        int[] arr = {5, 8, 6, 3, 9, 2, 1, 7};
        mergeSort(arr);
        System.out.println(Arrays.toString(arr));
    }

    public static void mergeSort(int[] arr) {
        int n = arr.length;
        int[] temp = new int[n];
        int pos = 0;            // 临时数组的下表
        int left1, left2;       // 左子数组边界
        int right1, right2;     // 右子数组边界

        for (int i = 1; i < n; i *= 2) {
            for (left1 = 0; left1 < n - i; left1 = right2) {
                // 设置子数组边界
                right1 = left2 = left1 + i;
                right2 = left2 + i;

                // 防止右边界越界
                right2 = right2 > n ? n : right2;

                pos = 0;
                while (left1 < left2 && right1 < right2) {
                    if (arr[left1] < arr[right1]) {
                        temp[pos++] = arr[left1++];
                    } else {
                        temp[pos++] = arr[right1++];
                    }
                }

                while (left1 < left2) {
                    arr[--right1] = arr[--left2];
                }

                // 将排好序的部分保存回数组
                while (pos > 0) {
                    arr[--right1] = temp[--pos];
                }
            }
        }
    }
}

File: code/Python/第13章 排序算法/13.2 归并排序/merge_sort_iteration.py
def merge_sort(lst):
    n = len(lst)
    temp = [0] * n
    pos = 0     # 临时列表的下标
    left1 = left2 = 0       # 左子数组边界
    right1 = right2 = 0     # 右子数组边界

    i = 1
    while i < n:
        left1 = 0
        while left1 < n - i:
            # 设置子数组边界
            right1 = left2 = left1 + i
            right2 = left2 + i

            # 防止右边界越界
            if right2 > n:
                right2 = n
            
            pos = 0
            while left1 < left2 and right1 < right2:
                if lst[left1] < lst[right1]:
                    temp[pos] = lst[left1]
                    pos += 1
                    left1 += 1
                else:
                    temp[pos] = lst[right1]
                    pos += 1
                    right1 += 1
            
            while left1 < left2:
                right1 -= 1
                left2 -= 1
                lst[right1] = lst[left2]

            # 将排好序的部分保存回数组
            while pos > 0:
                right1 -= 1
                pos -= 1
                lst[right1] = temp[pos]

            left1 = right2

        i *= 2

def main():
    lst = [5, 8, 6, 3, 9, 2, 1, 7]
    merge_sort(lst)
    print(lst)

if __name__ == "__main__":
    main()